build_runner/lib/src/build/asset_graph/nodes.dart

```diff
@@ -0,0 +1,239 @@
+    AssetId id,
+    void Function(AssetNodeBuilder) updates,
+  ) {
+    final node = get(id);
+    if (node == null) return null;
+    final updatedNode = node.rebuild(updates);
+    _nodes[id] = updatedNode;
+
+    if (node.inputs != updatedNode.inputs) {
+      _outputs = null;
+    }
+    if (node.isFile != updatedNode.isFile) {
+      _sortedFileIdsByPackage = null;
+    }
+
+    return updatedNode;
+  }
+
+  /// Adds [node] to the graph if it doesn't exist.
+  ///
+  /// Throws a [StateError] if it already exists in the graph.
+  ///
+  /// Returns the updated node: if it replaces an [AssetNode.missingSource] then
+  /// `outputs` and `primaryOutputs` are copied to it from that.
+  AssetNode add(AssetNode node) {
+    if (node.isFile) _sortedFileIdsByPackage = null;
+    final existing = get(node.id);
+    if (existing != null) {
+      if (existing.type == NodeType.missingSource) {
+        _nodes.remove(existing.id);
+        node = node.rebuild((b) {
+          b.primaryOutputs.addAll(existing.primaryOutputs);
+        });
+      } else {
+        throw StateError(
+          'Tried to add node ${node.id} to the asset graph but it already '
+          'exists.',
+        );
+      }
+    }
+    _nodes[node.id] = node;
+    if (node.inputs?.isNotEmpty ?? false) {
+      _outputs = null;
+    }
+
+    return node;
+  }
+
+  /// Removes [id] from the graph, if present.
+  void remove(AssetId id) {
+    final removed = _nodes.remove(id);
+    if (removed != null && removed.isFile) {
+      _sortedFileIdsByPackage = null;
+    }
+  }
+
+  /// All nodes in the graph.
+  Iterable<AssetNode> get allNodes => _nodes.values;
+
+  /// All IDs in `package` that refer to files and match.
+  ///
+  /// A node refers to a file if [AssetNode.isFile].
+  ///
+  /// If [glob] is passed, return only IDs for which the glob matches the path.
+  Iterable<AssetId> packageFileIds(String package, {Glob? glob}) {
+    _sortedFileIdsByPackage ??= _computeSortedFileIdsByPackage();
+    final list = _sortedFileIdsByPackage![package];
+    if (list == null) return const [];
+    if (glob == null) return list;
+
+    final simplePrefix = _simpleGlobPrefix(glob);
+    Iterable<AssetId> result = list;
+    if (simplePrefix.isNotEmpty) {
+      result = list.filterToPrefix(simplePrefix);
+    }
+
+    return result.where((id) => glob.matches(id.path));
+  }
+
+  /// The prefix of the [glob] pattern up to but not including the first special
+  /// character.
+  String _simpleGlobPrefix(Glob glob) {
+    final pattern = glob.pattern;
+    for (var i = 0; i != pattern.length; ++i) {
+      final char = pattern.codeUnitAt(i);
+      if (_specialGlobCharsCodeUnits.contains(char)) {
+        return pattern.substring(0, i);
+      }
+    }
+    return pattern;
+  }
+
+  /// Computes a `Map` from package names to sorted IDs of files in the package.
+  ///
+  /// An asset ID is a file if [AssetNode.isFile].
+  Map<String, List<AssetId>> _computeSortedFileIdsByPackage() {
```

Comment:
How often is this done? It seems to me we invalidate much more than we have to, but if it's only done, say, once in practise anyway it obviously doesn't matter --- though maybe a comment somewhere would be good.

Reply:
👍